src/cpp/cpp_typecheck_resolve.cpp

```diff
@@ -486,51 +486,44 @@ void cpp_typecheck_resolvet::disambiguate_functions(
     }
   }
 
-  identifiers.clear();
+  old_identifiers.clear();
 
   // put in the top ones
   if(!distance_map.empty())
   {
-    std::size_t distance=distance_map.begin()->first;
-
-    for(std::multimap<std::size_t, exprt>::const_iterator
-        it=distance_map.begin();
-        it!=distance_map.end() && it->first==distance;
-        it++)
-      identifiers.push_back(it->second);
+    auto range=distance_map.equal_range(distance_map.begin()->first);
+    for(auto it=range.first; it!=range.second; ++it)
+      old_identifiers.push_back(it->second);
   }
 
-  if(identifiers.size()>1 && fargs.in_use)
+  if(old_identifiers.size()>1 && fargs.in_use)
   {
     // try to further disambiguate functions
 
-    for(resolve_identifierst::iterator
-        it1=identifiers.begin();
-        it1!=identifiers.end();
+    for(resolve_identifierst::const_iterator
+        it1=old_identifiers.begin();
+        it1!=old_identifiers.end();
         it1++)
     {
+#if 0
+      std::cout << "I1: " << it1->get(ID_identifier) << std::endl;
+#endif
+
       if(it1->type().id()!=ID_code)
-        continue;
+      {
+	       identifiers.push_back(*it1);
+	       continue;
+      }
 
       const code_typet &f1=
         to_code_type(it1->type());
 
-      for(resolve_identifierst::iterator it2=
-          identifiers.begin();
-          it2!=identifiers.end();
-          ) // no it2++
+      for(resolve_identifierst::const_iterator it2=it1+1;
```

Comment:
Nit: giving these iterators meaningful names might make this code a bit easier to follow